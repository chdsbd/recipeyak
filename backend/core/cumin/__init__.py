import decimal
import enum
from collections import defaultdict
from dataclasses import dataclass, field
from decimal import Decimal
from enum import Enum
from typing import Dict, List, Optional, Sequence, Tuple, Union

from core.schedule.inflect import singularize


@dataclass
class UnhandledCase(Exception):
    case: "Unit"


@dataclass
class IncompatibleUnit(Exception):
    units: Tuple["Unit", "Unit"]


@enum.unique
class Unit(str, Enum):
    # mass/weight
    POUND = "POUND"
    OUNCE = "OUNCE"
    # metric
    GRAM = "GRAM"
    KILOGRAM = "KILOGRAM"

    # volume
    TEASPOON = "TEASPOON"
    TABLESPOON = "TABLESPOON"
    FLUID_OUNCE = "FLUID_OUNCE"
    CUP = "CUP"
    PINT = "PINT"
    QUART = "QUART"
    GALLON = "GALLON"
    # metric volume
    LITER = "LITER"
    MILLILITER = "MILLILITER"

    # covers "pinch", "sprinkle", "dash", "some", etc.
    SOME = "SOME"

    # for instance `1 cucumber`
    NONE = "NONE"

    # "1 bag", we don't parse bag so we just return UNKNOWN and let the caller
    # handle the result
    UNKNOWN = "UNKNOWN"

    def base_unit(self) -> "BaseUnit":
        if self in VOLUME:
            return BaseUnit.VOLUME
        if self in MASS:
            return BaseUnit.MASS
        if self == Unit.SOME:
            return BaseUnit.SOME
        if self == Unit.NONE:
            return BaseUnit.NONE
        if self == Unit.UNKNOWN:
            return BaseUnit.UNKNOWN
        raise UnhandledCase(case=self)

    def __lt__(self, other) -> bool:
        if self in VOLUME:
            return VOLUME[self] < VOLUME[other]
        if self in MASS:
            return MASS[self] < MASS[other]
        return False


TEASPOON_ML = Decimal(4.92892)
TABLESPOON_ML = 3 * TEASPOON_ML
FLUID_OUNCE_ML = 2 * TABLESPOON_ML
CUP_ML = 8 * FLUID_OUNCE_ML
PINT_ML = 2 * CUP_ML
QUART_ML = 2 * PINT_ML
GALLON_ML = 4 * QUART_ML

VOLUME: Dict[Unit, Decimal] = {
    Unit.MILLILITER: Decimal(1),
    Unit.TEASPOON: TEASPOON_ML,
    Unit.TABLESPOON: TABLESPOON_ML,
    Unit.FLUID_OUNCE: FLUID_OUNCE_ML,
    Unit.CUP: CUP_ML,
    Unit.PINT: PINT_ML,
    Unit.QUART: QUART_ML,
    Unit.LITER: Decimal(1000),
    Unit.GALLON: GALLON_ML,
}

OUNCE_GRAM = Decimal(28.34952)

MASS: Dict[Unit, Decimal] = {
    Unit.GRAM: Decimal(1),
    Unit.OUNCE: OUNCE_GRAM,
    Unit.POUND: 16 * OUNCE_GRAM,
    Unit.KILOGRAM: Decimal(1000),
}


@enum.unique
class BaseUnit(Enum):
    MASS = "MASS"
    VOLUME = "VOLUME"
    SOME = "SOME"
    NONE = "NONE"
    UNKNOWN = "UNKNOWN"


@dataclass
class Quantity:
    quantity: Decimal
    unit: Unit
    unknown_unit: Optional[str] = None

    def __add__(self, other: "Quantity") -> "Quantity":
        if self.unit == other.unit != Unit.UNKNOWN:
            return Quantity(quantity=self.quantity + other.quantity, unit=self.unit)
        if (
            self.unit == other.unit == Unit.UNKNOWN
            and self.unknown_unit == other.unknown_unit
        ):
            return Quantity(
                quantity=self.quantity + other.quantity,
                unit=Unit.UNKNOWN,
                unknown_unit=self.unknown_unit,
            )
        if self.unit == Unit.SOME:
            return other
        if other.unit == Unit.SOME:
            return self
        if self.unit.base_unit() == other.unit.base_unit():
            if self.unit == Unit.UNKNOWN:
                raise IncompatibleUnit(units=(self.unit, other.unit))
            unit_lookup = VOLUME if self.unit in VOLUME else MASS
            smallest_unit = self.unit if self.unit < other.unit else other.unit
            quantity = self.to_base_unit().quantity + other.to_base_unit().quantity
            return Quantity(
                quantity=quantity / unit_lookup[smallest_unit], unit=smallest_unit
            )
        raise IncompatibleUnit(units=(self.unit, other.unit))

    def __str__(self) -> str:
        if self.unit == Unit.SOME:
            return "some"
        if self.unit == Unit.NONE:
            return str(self.quantity)
        unit = str(self.unit)
        return f"{self.quantity} {unit}"

    def to_base_unit(self) -> "Quantity":
        if self.unit.base_unit() == BaseUnit.VOLUME:
            return Quantity(
                unit=Unit.MILLILITER, quantity=VOLUME[self.unit] * self.quantity
            )
        if self.unit.base_unit() == BaseUnit.MASS:
            return Quantity(unit=Unit.GRAM, quantity=MASS[self.unit] * self.quantity)
        return self


@dataclass
class Ingredient:
    quantity: str
    name: str
    description: str = ""


UNICODE_FRACTION_MAPPING = {
    "½": "1/2",
    "⅓": "1/3",
    "⅔": "2/3",
    "¼": "1/4",
    "¾": "3/4",
    "⅕": "1/5",
    "⅖": "2/5",
    "⅗": "3/5",
    "⅘": "4/5",
    "⅙": "1/6",
    "⅚": "5/6",
    "⅐": "1/7",
    "⅛": "1/8",
    "⅜": "3/8",
    "⅝": "5/8",
    "⅞": "7/8",
    "⅑": "1/9",
    "⅒": "1/10",
}


def unicode_fractions_to_ascii(s: str) -> str:
    """
    convert occurances of unicode fractions like `½` to their ascii
    equivalent `1/2`
    """

    output = ""
    for char in s:
        output += UNICODE_FRACTION_MAPPING.get(char, char)

    return output


quantity_chars = {c for c in " /.0123456789"}


def get_unit(val: str) -> Unit:
    val_cased = val.strip()
    val = val_cased.lower()
    if "cup" in val:
        return Unit.CUP
    if val == "kg":
        return Unit.KILOGRAM
    if "gram" in val:
        return Unit.GRAM
    if "gallon" in val:
        return Unit.GALLON
    if "ounce" in val or val == "oz":
        return Unit.OUNCE
    if "milliliter" in val:
        return Unit.MILLILITER
    if val_cased == "t" or "tsp" in val or "teaspoon" in val:
        return Unit.TEASPOON
    if val_cased == "T" or "tbs" in val or "tablespoon" in val:
        return Unit.TABLESPOON
    if "pinch" in val:
        return Unit.SOME
    if "lbs" in val or "lb" in val or "pound" in val:
        return Unit.POUND
    if "quart" in val:
        return Unit.QUART
    if "liter" in val:
        return Unit.LITER
    if val in {"some", "sprinkle", "dash"}:
        return Unit.SOME
    if val == "":
        return Unit.NONE
    return Unit.UNKNOWN


def fraction_to_decimal(val: str) -> Optional[Decimal]:
    try:
        total = Decimal(0)
        for v in val.strip().split(" "):
            if "/" in v:
                top, bot = v.split("/")
                total += Decimal(top) / Decimal(bot)
            else:
                total += Decimal(v)
        return total
    except decimal.InvalidOperation:
        return None


def max_quantity(quantity: str) -> str:
    """
    take the '4-5' medium button mushrooms and find the max, 5
    """
    if " to " in quantity:
        return quantity.split(" to ")[-1]
    return quantity.split("-")[-1]


def parse_quantity(val: str) -> Quantity:
    """
    handle "3 Tablespoon + 1 teaspoon" format
    """
    quantities = val.split("+")
    a = _parse_quantity(quantities[0])
    for q in quantities[1:]:
        a += _parse_quantity(q)
    return a


MALFORMED_UNITS = {"large", "medium", "small", "fresh"}


def _parse_quantity(val: str) -> Quantity:
    value = iter(unicode_fractions_to_ascii(max_quantity(val).strip()))

    quantity = ""
    unit_str = ""
    for c in value:
        if c in quantity_chars:
            quantity += c
        else:
            unit_str += c
            break

    for c in value:
        if c in quantity_chars:
            break
        # handle case of two units and quantities next to each other
        unit_str += c

    # strip out misplaced words, e.g., `1 large` `lemon` instead of `1` `large lemon`
    if unit_str in MALFORMED_UNITS:
        unit_str = ""

    unit = get_unit(unit_str)

    unknown_unit = None
    if unit == Unit.UNKNOWN:
        unknown_unit = unit_str

    return Quantity(
        quantity=fraction_to_decimal(quantity) or Decimal(1),
        unit=unit,
        unknown_unit=unknown_unit,
    )


def normalize_ingredient_name(*, name: str) -> str:
    return name.replace("-", " ").lower()


@dataclass
class IngredientItem:
    quantities: List[Quantity] = field(default_factory=list)
    category: Optional[str] = None


IngredientList = Dict[str, IngredientItem]


def combine_ingredients(ingredients: Sequence[Ingredient]) -> IngredientList:
    # being kind of dynamic with the types here so not the easiest on the eyes.
    ingredient_map: Dict[str, Dict[Union[BaseUnit, str, None], Quantity]] = defaultdict(
        dict
    )

    plural_name: Dict[str, str] = dict()

    for ingr in ingredients:
        normalized_name = normalize_ingredient_name(name=ingr.name)
        quantity = parse_quantity(ingr.quantity)
        base_unit = quantity.unit.base_unit()
        name = singularize(normalized_name)

        # keep track of whether an ingredient should be plural
        if name != normalized_name:
            plural_name[name] = normalized_name

        if base_unit == BaseUnit.UNKNOWN:
            # For each Unit.UNKNOWN, we treat the unknown_unit, as a unique
            # base value.
            if quantity.unknown_unit not in ingredient_map[name]:
                ingredient_map[name][quantity.unknown_unit] = quantity
            else:
                ingredient_map[name][quantity.unknown_unit] += quantity
        elif name not in ingredient_map:
            ingredient_map[name][base_unit] = quantity
        else:
            base_unit_quantity = ingredient_map[name].get(base_unit)
            if base_unit_quantity is None:
                ingredient_map[name][base_unit] = quantity
            else:
                ingredient_map[name][base_unit] = base_unit_quantity + quantity

    output: IngredientList = defaultdict(IngredientItem)

    for ingre_name, unit_to_quantity in ingredient_map.items():
        name = plural_name.get(ingre_name, ingre_name)
        output[name].quantities += unit_to_quantity.values()

    return output
