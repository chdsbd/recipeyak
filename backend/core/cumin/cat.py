DEPARTMENT_MAPPING = {
    "produce": {
        "basil",
        "coconut",
        "plum",
        "apricots",
        "sage",
        "tarragon",
        "romaine",
        "habanero chile",
        "snow pea",
        "spring mix",
        "dates",
        "prunes",
        "corn",
        "tomatillos",
        "salad",
        "pico de gallo",
        "peas",
        "bean sprouts",
        "passion fruit",
        "mushroom",
        "herbs",
        "pasilla chilies",
        "mulato chiles",
        "haricot verts",
        "haricots verts",
        "greens",
        "pomegranate",
        "edamame",
        "rhubarb",
        "zucchini",
        "watermelon",
        "brussel sprouts",
        "brussels sprouts",
        "brusel sprouts",
        "dill",
        "beets",
        "broccoli",
        "berries",
        "leeks",
        "ancho chiles",
        "chipotle chiles",
        "pequin chile",
        "cascabel chile",
        "arbor chile",
        "chile negro",
        "chile ancho",
        "cauliflower",
        "fennel",
        "cranberries",
        "scallion",
        "bok choy",
        "turnip",
        "red chili",
        "tofu",
        "potato",
        "pear",
        "strawberries",
        "strawberris",
        "strawberry",
        "grapefruit",
        "serrano chile",
        "thai chile",
        "green chillies",
        "green chiles",
        "carrot",
        "chive",
        "avocado",
        "watercress",
        "cabbage",
        "scallions",
        "lime",
        "thyme",
        "radish",
        "mushroom",
        "lemon",
        "tomato",
        "lettuce",
        "shallot",
        "rosemary",
        "apple",
        "banana",
        "kale",
        "cilantro",
        "asparagus",
        "spinach",
        "orange",
        "arugula",
        "jalapeno",
        "jalapeño",
        "bell pepper",
        "onion",
        "parsley",
        "parsely",
        "mint",
        "garlic",
        "ginger",
        "lemon",
        "clementine",
        "celery",
        "oregano",
        "cucumber",
    },
    "meat": {
        "chicken",
        "beef",
        "pork",
        "bacon",
        "sausage",
        "salmon",
        "kielbasa",
        "merguez",
        "chuck-eye roast",
        "pancetta",
        "oxtails",
        "ham",
        "chuck roast",
        "round roast",
        "bratwurst",
        "steak",
        "prawns",
        "turkey",
        "veal",
        "sirloin",
        "chorizo",
        "short ribs",
        "shortribs",
        "lamb",
        "fish",
        "brisket",
    },
    "cheese": {
        "cheese",
        "cheddar",
        "colby jack",
        "provolone",
        "queso fresco",
        "mascarpone",
        "fontina",
        "pecorino",
        "feta",
        "parmesan",
        "gruyère",
        "gruyere",
        "mozzarella",
    },
    "dairy": {
        "egg",
        "milk",
        "cream",
        "ricotta",
        "yogurt",
        "yoghurt",
        "butter",
        "half-and-half",
        "creme",
        "creme fraiche",
        "crème fraîche",
    },
    "baking": {
        "sugar",
        "flour",
        "chocolate",
        "choclate",
        "poppy seeds",
        "molasses",
        "lard",
        "ghee",
        "gee",
        "cocoa",
        "food coloring",
        "cocoa powder",
        "coca powder",
        "baking soda",
        "baking powder",
    },
    "frozen": {
        "bag potstickers",
        "potstickers",
        "phyllo dough",
        "phyllo pastry",
        "puff pastry",
        "pie dough",
        "pie crust",
        "pie shell",
    },
    "bread": {
        "bread",
        "panettone",
        "naan",
        "burger bun",
        "pita",
        "rolls",
        "brioche",
        "baguette",
        "tortillas",
        "sub roll",
        "english muffins",
        "bagel",
        "hot dog buns",
        "pizza dough",
    },
    "canned": {
        "anchovy",
        "fruit preserves",
        "chickpeas",
        "beans",
        "sardines",
        "pumpkin puree",
        "pumpkin purée",
        "marinara sauce",
        "tuna",
        "anchovies",
        "adobo",
        "stock",
    },
    # kind of the misc category
    "condiments": {
        "tamari",
        "pesto",
        "gochujang",
        "salsa",
        "pickle",
        "pizza sauce",
        "mayo",
        "mayonnaise",
        "mango chutney",
        "marshmallows",
        "ketchup",
        "ketcup",
        "oil",
        "honey",
        "dulce de leche",
        "hot sauce",
        "fish sauce",
        "capers",
        "vinegar",
        "soy sauce",
        "tamarind",
        "tahini",
        "tzatziki",
        "sprinkles",
        "sesame paste",
        "mustard",
        "seaweed",
        "jam",
        "maple syrup",
        "marmite",
        "sriracha",
        "ya cai",
        "yacai",
        "curry paste",
        "worcestershire sauce",
        "bean paste",
        "almond paste",
        "agave nectar",
        "olives",
    },
    "dry goods": {
        "powdered msg",
        "tortilla chips",
        "quinoa",
        "raisins",
        "masa harina",
        "polenta",
        "panko",
        "orzo",
        "oatmeal",
        "oats",
        "lentils",
        "ladyfingers",
        "nilla wafers",
        "granola",
        "fritos",
        "bulgar",
        "bulgur",
        "millet",
        "flax seed",
        "coffee",
        "espresso",
        "chia",
        "couscous",
        "gelatin",
        "bati masala",
        "amaranth",
        "barley",
        "farro",
        "ramen",
        "fettuccine",
        "macaroni",
        "fusilli",
        "rotini",
        "penne",
        "rigatoni",
        "yeast",
        "rice",
        "pasta",
        "tapioca",
        "cornstarch",
        "noodles",
        "spaghetti",
    },
    "alcohol": {
        "wine",
        "beer",
        "rum",
        "vodka",
        "vermouth",
        "amaretto",
        "mirin",
        "bourbon",
        "brandy",
        "scotch",
        "cognac",
        "sherry",
        "guinness",
        "stout",
        "madeira",
        "marsala",
    },
    "spices": {
        "cinnamon",
        "curry powder",
        "sumac",
        "paprika",
        "salt",
        "caraway seeds",
        "bati masala",
        "five-spice powder",
        "allspice",
        "pepper",
        "bay leaves",
        "miso",
        "cloves",
        "ground clove",
        "red chile flakes",
        "chile flakes",
        "chili flakes",
        "chili powder",
        "chilli powder",
        "chile powder",
        "cumin",
        "coriander",
        "corainder",
        "vanilla",
        "whole clove",
        "nutmeg",
        "turmeric",
        "sesame seed",
        "mustard seed",
        "dried chilies",
        "dried chiles",
        "fenugreek",
        "ground chiles",
        "baharat",
        "seasoning",
        "star anise",
        "marjoram",
        "bay leaf",
        "almond extract",
        "cardamom",
        "cayenne",
        "cream of tartar",
        "garam masala",
        "herbes de provence",
        "saffron",
        "niger seed",
        "nigella seed",
        "peperoncino flakes",
    },
    "nuts": {
        "pecans",
        "pecan",
        "pine nuts",
        "peanuts",
        "walnut",
        "almonds",
        "nuts",
        "cashews",
        "pistachios",
    },
    "other": {"cedar shakes", "water", "ice", "gloves", "skewers",},
}


def category(ingredient: str) -> str:
    for cat, keywords in DEPARTMENT_MAPPING.items():
        for k in keywords:
            if k in ingredient:
                return cat
    return "unknown"
