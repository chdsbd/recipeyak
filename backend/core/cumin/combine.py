from __future__ import annotations

from collections import defaultdict
from dataclasses import dataclass, field
from typing import Dict, Sequence

from core.cumin.quantity import BaseUnit, Quantity, parse_quantity
from core.schedule.inflect import singularize


@dataclass
class Ingredient:
    quantity: str
    name: str
    description: str = ""


def normalize_ingredient_name(*, name: str) -> str:
    return name.replace("-", " ").lower()


@dataclass
class IngredientItem:
    quantities: list[Quantity] = field(default_factory=list)
    category: str | None = None


IngredientList = Dict[str, IngredientItem]


def combine_ingredients(ingredients: Sequence[Ingredient]) -> IngredientList:
    # being kind of dynamic with the types here so not the easiest on the eyes.
    ingredient_map: dict[str, dict[BaseUnit | str | None, Quantity]] = defaultdict(dict)

    plural_name: dict[str, str] = dict()

    for ingr in ingredients:
        normalized_name = normalize_ingredient_name(name=ingr.name)
        quantity = parse_quantity(ingr.quantity)
        base_unit = quantity.unit.base_unit()
        name = singularize(normalized_name)

        # keep track of whether an ingredient should be plural
        if name != normalized_name:
            plural_name[name] = normalized_name

        if base_unit == BaseUnit.UNKNOWN:
            # For each Unit.UNKNOWN, we treat the unknown_unit, as a unique
            # base value.
            if quantity.unknown_unit not in ingredient_map[name]:
                ingredient_map[name][quantity.unknown_unit] = quantity
            else:
                ingredient_map[name][quantity.unknown_unit] += quantity
        elif name not in ingredient_map:
            ingredient_map[name][base_unit] = quantity
        else:
            base_unit_quantity = ingredient_map[name].get(base_unit)
            if base_unit_quantity is None:
                ingredient_map[name][base_unit] = quantity
            else:
                ingredient_map[name][base_unit] = base_unit_quantity + quantity

    output: IngredientList = defaultdict(IngredientItem)

    for ingre_name, unit_to_quantity in ingredient_map.items():
        name = plural_name.get(ingre_name, ingre_name)
        output[name].quantities += unit_to_quantity.values()

    return output
