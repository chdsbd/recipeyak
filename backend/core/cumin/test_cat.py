import csv
from collections import defaultdict
from pathlib import Path
from textwrap import dedent

from core.cumin.cat import category, create_trie, search


def test_categorize_ingredients() -> None:
    """
    Ensure we have a category set for each ingredient
    """
    out = defaultdict(set)
    with (Path(".") / "ingredients.csv").open() as f:
        reader = csv.DictReader(f)
        for row in reader:
            item = row["name"]
            cat = category(item.lower())
            out[cat].add(item)

    expected_unknown = {
        "for a 9-inch double crust pie (see other recipe)",
        "thinly sliced",
    }

    out["unknown"] -= expected_unknown

    assert not out["unknown"], f'remaining {len(out["unknown"])}'


def test_categorize_ingredient_test_cases(snapshot) -> None:
    cases = dedent(
        """
    (13-ounce) can unsweetened coconut milk
    (14.5-ounce) can whole or crushed tomatoes
    (15-ounce) cans chickpeas
    (6-inch) corn tortillas
    1 lime
    1-inch piece ginger
    1/4 ounce active dry yeast
    14 oz cans of diced tomatoes
    14.5-ounce can whole peeled tomatoes
    15-ounce can chickpeas
    15-ounce container ricotta cheese
    28 ounce can San Marzano tomatoes
    28-ounce can whole tomatoes
    3-to-3.5-pound butternut squash
    4-ounce milk chocolate bar
    Asian fish sauce
    Chinese rice wine
    Grated Parmigiano-Reggiano
    all-butter puff pastry
    almond butter
    almond paste
    apple cider vinegar
    baby bok choy
    beef broth
    beef or chicken broth
    beef stock
    best-quality apricot jam or orange marmalade
    bulk hot or sweet breakfast sausage or Italian sausage
    butternut squash
    butternut squash (1 small squash)
    chicken stock
    chinese egg noodles
    coconut extract
    coconut oil
    coconut rum or dark rum
    coconut sugar or substitute brown sugar
    corn torillas
    corn tortilla chips
    corn tortillas
    cornmeal
    cornstarch
    cream of coconut
    cream of tartar
    creamy unsweetened peanut butter
    crystallized ginger
    cubanelle peppers
    dark corn syrup
    diced canned tomatoes
    dill pickle chips
    doubanjiang (Sichuanese fermented chile bean paste; see Note)
    dried green split peas
    dry mustard powder
    eggplant
    finely chopped cornichons or small kosher dill pickles
    finely grated ginger
    fish sauce
    graham cracker crumbs
    ground ginger
    hamburger buns
    haricot vert fins or green beans
    harissa or tomato paste
    hot italian sausage
    large poblano pepper
    lemon extract
    light corn syrup
    loaf sweet egg bread like challah or brioche
    low sodium chicken stock
    low-sodium chicken broth
    malted milk powder
    medium-coarse yellow cornmeal
    """
    ).splitlines()

    assert sorted((w, category(w)) for w in cases) == snapshot()


def test_trie() -> None:
    mapping = {
        "spices": {
            "red chile flakes",
            "chile powder",
        }
    }

    trie = create_trie(mapping)
    assert search("red chile flakes", trie=trie)
    assert search("red chile powder", trie=trie)
