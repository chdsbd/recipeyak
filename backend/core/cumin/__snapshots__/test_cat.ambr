# name: test_categorize_ingredient_test_cases
  <class 'list'> [
    <class 'tuple'> (
      '',
      'unknown',
    ),
    <class 'tuple'> (
      '(13-ounce) can unsweetened coconut milk',
      'produce',
    ),
    <class 'tuple'> (
      '(14.5-ounce) can whole or crushed tomatoes',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '(15-ounce) cans chickpeas',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '(6-inch) corn tortillas',
      'bread',
    ),
    <class 'tuple'> (
      '1 lime',
      'produce',
    ),
    <class 'tuple'> (
      '1-inch piece ginger',
      'produce',
    ),
    <class 'tuple'> (
      '1/4 ounce active dry yeast',
      'dry goods',
    ),
    <class 'tuple'> (
      '14 oz cans of diced tomatoes',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '14.5-ounce can whole peeled tomatoes',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '15-ounce can chickpeas',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '15-ounce container ricotta cheese',
      'dairy',
    ),
    <class 'tuple'> (
      '28 ounce can San Marzano tomatoes',
      'produce',
    ),
    <class 'tuple'> (
      '28-ounce can whole tomatoes',
      'canned & packaged',
    ),
    <class 'tuple'> (
      '3-to-3.5-pound butternut squash',
      'produce',
    ),
    <class 'tuple'> (
      '4-ounce milk chocolate bar',
      'baking',
    ),
    <class 'tuple'> (
      'Asian fish sauce',
      'condiments',
    ),
    <class 'tuple'> (
      'Chinese rice wine',
      'dry goods',
    ),
    <class 'tuple'> (
      'Grated Parmigiano-Reggiano',
      'cheese',
    ),
    <class 'tuple'> (
      'all-butter puff pastry',
      'frozen',
    ),
    <class 'tuple'> (
      'almond butter',
      'condiments',
    ),
    <class 'tuple'> (
      'almond paste',
      'condiments',
    ),
    <class 'tuple'> (
      'apple cider vinegar',
      'condiments',
    ),
    <class 'tuple'> (
      'baby bok choy',
      'produce',
    ),
    <class 'tuple'> (
      'beef broth',
      'meat',
    ),
    <class 'tuple'> (
      'beef or chicken broth',
      'meat',
    ),
    <class 'tuple'> (
      'beef stock',
      'canned & packaged',
    ),
    <class 'tuple'> (
      'best-quality apricot jam or orange marmalade',
      'produce',
    ),
    <class 'tuple'> (
      'bulk hot or sweet breakfast sausage or Italian sausage',
      'meat',
    ),
    <class 'tuple'> (
      'butternut squash',
      'produce',
    ),
    <class 'tuple'> (
      'butternut squash (1 small squash)',
      'produce',
    ),
    <class 'tuple'> (
      'chicken stock',
      'canned & packaged',
    ),
    <class 'tuple'> (
      'chinese egg noodles',
      'dry goods',
    ),
    <class 'tuple'> (
      'coconut extract',
      'produce',
    ),
    <class 'tuple'> (
      'coconut oil',
      'produce',
    ),
    <class 'tuple'> (
      'coconut rum or dark rum',
      'produce',
    ),
    <class 'tuple'> (
      'coconut sugar or substitute brown sugar',
      'produce',
    ),
    <class 'tuple'> (
      'corn torillas',
      'produce',
    ),
    <class 'tuple'> (
      'corn tortilla chips',
      'dry goods',
    ),
    <class 'tuple'> (
      'corn tortillas',
      'bread',
    ),
    <class 'tuple'> (
      'cornmeal',
      'dry goods',
    ),
    <class 'tuple'> (
      'cornstarch',
      'dry goods',
    ),
    <class 'tuple'> (
      'cream of coconut',
      'canned & packaged',
    ),
    <class 'tuple'> (
      'cream of tartar',
      'spices',
    ),
    <class 'tuple'> (
      'creamy unsweetened peanut butter',
      'condiments',
    ),
    <class 'tuple'> (
      'crystallized ginger',
      'produce',
    ),
    <class 'tuple'> (
      'cubanelle peppers',
      'produce',
    ),
    <class 'tuple'> (
      'dark corn syrup',
      'produce',
    ),
    <class 'tuple'> (
      'diced canned tomatoes',
      'canned & packaged',
    ),
    <class 'tuple'> (
      'dill pickle chips',
      'condiments',
    ),
    <class 'tuple'> (
      'doubanjiang (Sichuanese fermented chile bean paste; see Note)',
      'condiments',
    ),
    <class 'tuple'> (
      'dried green split peas',
      'dry goods',
    ),
    <class 'tuple'> (
      'dry mustard powder',
      'condiments',
    ),
    <class 'tuple'> (
      'eggplant',
      'produce',
    ),
    <class 'tuple'> (
      'finely chopped cornichons or small kosher dill pickles',
      'condiments',
    ),
    <class 'tuple'> (
      'finely grated ginger',
      'produce',
    ),
    <class 'tuple'> (
      'fish sauce',
      'condiments',
    ),
    <class 'tuple'> (
      'graham cracker crumbs',
      'dry goods',
    ),
    <class 'tuple'> (
      'ground ginger',
      'produce',
    ),
    <class 'tuple'> (
      'hamburger buns',
      'bread',
    ),
    <class 'tuple'> (
      'haricot vert fins or green beans',
      'produce',
    ),
    <class 'tuple'> (
      'harissa or tomato paste',
      'produce',
    ),
    <class 'tuple'> (
      'hot italian sausage',
      'meat',
    ),
    <class 'tuple'> (
      'large poblano pepper',
      'spices',
    ),
    <class 'tuple'> (
      'lemon extract',
      'produce',
    ),
    <class 'tuple'> (
      'light corn syrup',
      'produce',
    ),
    <class 'tuple'> (
      'loaf sweet egg bread like challah or brioche',
      'bread',
    ),
    <class 'tuple'> (
      'low sodium chicken stock',
      'canned & packaged',
    ),
    <class 'tuple'> (
      'low-sodium chicken broth',
      'meat',
    ),
    <class 'tuple'> (
      'malted milk powder',
      'dairy',
    ),
    <class 'tuple'> (
      'medium-coarse yellow cornmeal',
      'dry goods',
    ),
  ]
---
